cmd/serve.go

```diff
@@ -31,37 +32,61 @@ var (
 	bindAddr          string
 	ignoredUserAgents []string
 	cacheDuration     time.Duration
+	refreshInterval   time.Duration
 )
 
 var serveCmd = &cobra.Command{
 	Use:   "serve",
 	Short: "Serves a webserver with an HTML listing of all images and tags in a v2 registry",
-	RunE: func(cmd *cobra.Command, args []string) error {
+	Run: func(cmd *cobra.Command, args []string) {
 		ctx := cmd.Context()
 		log := logger.FromContext(ctx)
 		log.Info("starting server",
 			slog.Duration("cache-duration", cacheDuration),
 			slog.String("bind-addr", bindAddr),
 			slog.Any("ignored-user-agents", ignoredUserAgents),
+			slog.Any("refresh-interval", refreshInterval),
 		)
 
 		client := registry.New(rootCfg)
+		asyncClient := async.New(client, refreshInterval)
 
-		filler := filler.New(client, rootCfg.RegistryHostname, "/")
+		filler := filler.New(asyncClient, rootCfg.RegistryHostname, "/")
 
-		regServer := staticreg.New(client, filler, rootCfg.RegistryHostname)
+		regServer := staticreg.New(asyncClient, filler, rootCfg.RegistryHostname)
 		srv, err := server.New(bindAddr, regServer, log, cacheDuration, ignoredUserAgents)
 		if err != nil {
-			return err
+			slog.Error("error creating server", logger.ErrAttr(err))
+			return
 		}
 
-		return srv.Start()
+		errCh := make(chan error, 1)
+		go func() {
+			errCh <- srv.Start()
+		}()
+
+		go func() {
+			errCh <- asyncClient.Start(ctx)
+		}()
+
+		select {
+		case <-ctx.Done():
+			return
+		case err := <-errCh:
+			if err == nil {
+				slog.Error("operations exited unexpectedly")
+				return
+			}
+			slog.Error("unexpected error", logger.ErrAttr(err))
+			return
+		}
```

Comment:
Also: any specific reason for using an error channel rather than `errgroup` (for example)?

Reply:
Those two goroutines in this context are long running tasks that should complete only on error or if the ctx gets canceled, whenever an error happens or one of the goroutines stops the program should also stop.

Those goroutines are thought in a way that they only error if they can't recover from it and if they can't I think that staticreg should just stop so that the system operating it, for example, a k8s cluster, sees that staticreg is restarting and can notify the operator.

Re `errgroup`: I'm just used to handling goroutine cancelation with an errCh + a select as I always felt it gave me more control and its easier to extend (not that it is a good thing :D).

Reply:
@alberto-miranda I've been thinking about this and I tried your proposed solution, that also made me go more in depth with my current implementation and found a few other flaws. Please take a look again!

Reply:
See [`3d58b00` (#16)](https://github.com/seqeralabs/staticreg/pull/16/commits/3d58b001d97d686eeee30f50ff10119d9d80732f)

Reply:
The current implementation LGTM, thanks. 

PS: I noticed that under normal conditions the program would just stop, but the reason I mentioned it was because I recently lost a couple hours due to a weird situation with dangling goroutines that only happened while unit testing some functions that internally created those goroutines. So, if you ever decided to test the `serveCmd.Run()` function outside of a program context, you might have been hit with something similar.

Reply:
I'm glad you commented, without having to re-reason this I would've left out a few bugs. Thank you :pray: